### ISSUE TITLE
Prepared statements

### ISSUE THREAD

**Issue by tgriesser** [tgriesser/knex#802]
Discussion around providing an API for normalizing true database prepared statements where possible (pg, mysql2, oracle?).

Issues:
- prepared statements are executed per-connection
- does not make as much sense with a dynamically constructed query, unless a cache/prepare api is exposed similar to #798

Prior: #718, #305

/cc @mibi-r

**Comment by myndzi** [comment]
The nature of a prepared query lends itself to the form:

    var prepared = buildPreparedQuery(...);
    ...
    prepared(arguments);

so that does tie in nicely with #798 (and depending on API choices might give it to you for free)

The tricky bit seems to be keeping track of and/or "upgrading" connections from the resource pool that can execute the query. A few modifications to pool2's "capabilities" option may provide a way to accomplish this fairly cleanly. 

Opened an issue over here to discuss this possibility: https://github.com/myndzi/pool2/issues/5

**Comment by taoeffect** [comment]
[node-sql](https://github.com/brianc/node-sql) correctly constructs parameterized queries for many dialects.

**Comment by elhigu** [comment]
@taoeffect also Knex builds internally parametrized queries for many dialects... could you explain a bit more how your comment helps to resolve this issue?

**Comment by taoeffect** [comment]
> also Knex builds internally parametrized queries for many dialects...

I did not see that while looking through the source...

> could you explain a bit more how your comment helps to resolve this issue?

Sure, it's a library that solves this problem, so taking a look at how it does it could be useful.

**Comment by taoeffect** [comment]
> I did not see that while looking through the source...

OK, sorry that was my mistake:

- I had searched for `'?'`  on GitHub and it revealed zero results, but I guess that's just a problem with GitHub search
- I stumbled upon this issue and I'd read it as Knex simply not supporting parameterized queries

Indeed, it's unclear to me what the extent of support for parameterized queries is in Knex.

I now see, looking at the unit tests, that it does seem to support them. However [these lines are not clear to me](https://github.com/tgriesser/knex/blob/dfca38da339922f8ad8f52087ca5e6c0dcdab081/test/unit/query/builder.js#L1810-L1843):

Why is it that here it shows parameterized queries:

```js
  it("multiple inserts", function() {
    testsql(qb().from('users').insert([{email: 'foo', name: 'taylor'}, {email: 'bar', name: 'dayle'}]), {
      mysql: {
        sql: 'insert into `users` (`email`, `name`) values (?, ?), (?, ?)',
        bindings: ['foo', 'taylor', 'bar', 'dayle']
      },
      sqlite3: {
        sql: 'insert into "users" ("email", "name") select ? as "email", ? as "name" union all select ? as "email", ? as "name"',
        bindings: ['foo', 'taylor', 'bar', 'dayle']
      },
```

And yet for a nearly identical query after it, it does not?

```js
it("multiple inserts with partly undefined keys client with configuration nullAsDefault: true", function() {
    testquery(qb().from('users').insert([{email: 'foo', name: 'taylor'}, {name: 'dayle'}]), {
      mysql: "insert into `users` (`email`, `name`) values ('foo', 'taylor'), (NULL, 'dayle')",
      sqlite3: 'insert into "users" ("email", "name") select \'foo\' as "email", \'taylor\' as "name" union all select NULL as "email", \'dayle\' as "name"',
```

**Comment by elhigu** [comment]
@taoeffect Actually knex supports 2 different ways of creating queries, knex can do the escaping by it self and then generate string which contains all parameters already replaced. But usually when knex just runs query it passes parametrized query and parameter bindings separately to DB driver which executes query. 

If you call knex's `toString()` or `toQuery()` methods, querybuilder constructs one string where all the bindings are evaluated to one string.

```js
let knex = require('knex')({client: 'pg'})

knex('table').where('id', 1).toString();
// 'select * from "table" where "id" = \'1\''
```

If you call knex's `toSQL()` query builder constructs parametrized SQL string and bindings, which are passed to driver.

```js
let knex = require('knex')({client: 'pg'})
knex('table').where('id', 1).toSQL();
// { method: 'select',
//    options: {},
//    bindings: [ 1 ],
//    sql: 'select * from "table" where "id" = ?' }
```

You can check the code for actually executing the query for postgres here 
https://github.com/tgriesser/knex/blob/master/src/dialects/postgres/index.js#L147

Few lines earlier there is call which converts `?` bindings to postgres format `$1`, `$2`... etc.

You can always check how knex is sending query by calling `toSQL()` and checking if it is parametrized or not...
 
I was trying to write some exotic query which is using knex.raw inside subquery, which is used by main query and wasn't able to make knex to lower any values to strings by itself

```js
subQuery = knex('table2').where('id', knex.raw('(select ?? from ?? where ?? = ?)', ['id', 'table3', 'name', knex.raw('LOWER(?)', ['john'])] ))
knex('table5').whereExists(subQuery).whereIn('id', [1, knex.raw('?', [2])]).toSQL()
{ method: 'select',
  options: {},
  bindings: [ 'john', 1, 2 ],
  sql: 'select * from "table5" where exists (select * from "table2" where "id" = (select "id" from "table3" where "name" = LOWER(?))) and "id" in (?, ?)' }
```

During the years I've been using knex I have never heard any bug reports of sql injection either so I'm pretty confident of that part of knex working well :)

**Comment by taoeffect** [comment]
> knex can do the escaping by it self and then generate string which contains all parameters already replaced.

So that's not good, IMO. It should always use parameterized queries unless someone is generating an `.sql` file database dump.

> But usually when knex just runs query it passes parametrized query and parameter bindings separately to DB driver which executes query.

That word "usually" is what I find confusing and makes me nervous. When does it and when doesn't it?

I still don't understand, for example, what the difference is [here](https://github.com/tgriesser/knex/blob/dfca38da339922f8ad8f52087ca5e6c0dcdab081/test/unit/query/builder.js#L1810-L1843). Why does one generate parameterized queries while the other doesn't?

That sort of uncertainty is what makes me hold off on Knex and the ORMs that use it. There's just no good reason to rely on inefficient and potentially insecure string escaping.

**Comment by taybin** [comment]
Any progress on support prepared statements?

**Comment by elhigu** [comment]
No and agreed with @taybin that this issue is too vague to be useful. If someone likes to have support for something regarding prepared statements (I'm not exactly sure what kind of support) new issue with API proposal + investigation how to support them in sqlite, mysql, postgres, mssql and oracle would be necessary. 

Closing this since information heris not really useful for anything.

@taybin I haven't heard that anyone would have been working on anything like this. Might be even necessary to have better control of connections which are use for different queries... not sure though.

**Comment by kswope** [comment]
Why did this issue go nowhere?  This seems like an important improvement, at least in the case of mysql.  Not only do you get the safety of server-side prepared statements, but no need for repeated client side escaping and transmission of SQL.   

You can see the difference in the mysqld logs between using mysql2 execute() and knex, which is probably just using query().

mysql2
```
2019-04-09T00:00:51.657686Z       118 Connect   root@localhost on test_development using TCP/IP                       
2019-04-09T00:00:51.659328Z       118 Prepare   select * from users where id=?                                        
2019-04-09T00:00:51.661270Z       118 Execute   select * from users where id=1                                        
2019-04-09T00:00:51.665140Z       118 Execute   select * from users where id=1   
``` 

knex using mysql2
```
2019-04-09T00:01:43.253314Z       119 Connect   root@localhost on test_development using TCP/IP                       
2019-04-09T00:01:43.261132Z       119 Query     select * from `users` where `id` = 1                                  
2019-04-09T00:01:43.266717Z       119 Query     select * from `users` where `id` = 1 
```

**Comment by elhigu** [comment]
> Why did this issue go nowhere? 

Nobody has needed the feature so badly that they would have implemented it. Also because this issue is mixing couple of separate things in its comments.

> Not only do you get the safety of server-side prepared statements, but no need for repeated client side escaping and transmission of SQL.

> You can see the difference in the mysqld logs between using mysql2 execute() and knex, which is probably just using query().

Original issue was about having real prepared statements, which are sent to session  and then being able to be executed multiple times with different parameters and having that feature available in a crossdb compatible manner.

For example pg-driver always creates first temporary prepared statement and only after that runs the query with parameters.

Also looks like that you are right that mysql2 client always escapes parameter bindings when .query(sql, bindings, ...) method is used https://github.com/mysqljs/mysql#escaping-query-values 

So it would be probably good idea to change that to use automatic cache/creation https://github.com/sidorares/node-mysql2/blob/master/documentation/Prepared-Statements.md#automatic-creation-cached-and-re-used-by-connection as you mentioned.

You may also check if mysql driver does something in a different way. In our stress tests mysql driver was a bit more stable than mysql2 when connections are closed unexpectedly.

@kswope Please open proper feature request for that. This issue was not about the thing you are asking for and is also too messed up with various different conversations.

**Comment by kswope** [comment]
> Original issue was about having real prepared statements, which are sent to session and then being able to be executed multiple times with different parameters and having that feature available in a crossdb compatible manner.

This was my point, that's why I included the mysql logs and not toString() which is what everybody seems to do and then gets reprimanded :)

Although I don't think I explained it fully.  In node-mysql2  query() and execute() seem to be interchangeable.  But with execute() you get real server-side prepared statements (subject of this issue).  

The first log proves server-side prepared statements, and I had to use node-mysql2 directly to call execute(). 

 knex is the second log entry, sending over the whole query, which is the exact same thing I'd get with the first example if I had swapped out execute() for query(), so I assume knex is using query().

Here's the thing, node-mysql (without the 2), has query(), but doesn't have execute().

I'm just thinking, if knex recognizes that node-mysql2 is being used instead of node-mysql, it could bypass all the client side escaping and just call execute() and let the driver take over and get server-side prepares for free.

**Comment by elhigu** [comment]
> Although I don't think I explained it fully. In node-mysql2 query() and execute() seem to be interchangeable. But with execute() you get real server-side prepared statements (subject of this issue).

You did explain it just fine in the first place, thats why I said you are right and pasted a link to documentation, which explains what mysql does with `.query(sql, args, cb)` (does escaping for values and creates a string). `.execute` is clearly preferable way to go and as a bonus it has internally built cache for those prepared statements (IIRC pg-driver doesn't cache those).

> Here's the thing, node-mysql (without the 2), has query(), but doesn't have execute().

Thats good to know, so we need to change only mysql2 drivers implementation how to send queries to the driver.

Anyways I'm really surprised if mysql driver doesn't offer any way to achieve similar behaviour (except of caching those statements). It could be tried out if it is easy and robust to implement. Would need some performance testing though to be able to verify that creating always prepared statements without caching them doesn't cause huge overhead to DB server side.

> I'm just thinking, if knex recognizes that node-mysql2 is being used instead of node-mysql, it could bypass all the client side escaping and just call execute() and let the driver take over and get server-side prepares for free.

Knex is not doing any of those escapings, it all happends inside mysql dirver and yes it is easy change to do and existing tests should be enough to verify that it works. However just to be sure I would like to mark it as breaking change (it is a pretty fundamental change so better be sae than sorry) even that strictly speaking it is not part of public API of knex.

As I said before, please open a separate feature request for that. We are talking about different issues  here than original post did.

**Comment by elhigu** [comment]
Locked conversation in here to force everyone to open clean and properly scoped feature requests for prepared statements / parameter bindings related issues.
